lib/pstore.rb

```diff
@@ -10,87 +10,290 @@
+# User code may later read values back from the data store
+# or even update data, as needed.
 #
 # The transactional behavior ensures that any changes succeed or fail together.
-# This can be used to ensure that the data store is not left in a transitory
-# state, where some values were updated but others were not.
+# This can be used to ensure that the data store is not left in a transitory state,
+# where some values were updated but others were not.
+#
+# Behind the scenes, Ruby objects are stored to the data store file with Marshal.
+# That carries the usual limitations. Proc objects cannot be marshalled,
+# for example.
+#
+# There are three key terms here (details at the links):
+#
+# - {Store}[rdoc-ref:PStore@The+Store]: a store is an instance of \PStore.
+# - {Roots}[rdoc-ref:PStore@Roots]: the store is hash-like;
+#   each root is a key for a stored object.
+# - {Transactions}[rdoc-ref:PStore@Transactions]: each transaction is a collection
+#   of prospective changes to the store;
+#   a transaction is defined in the block given with a call
+#   to PStore#transaction.
+#
+# == About the Examples
+#
+# All examples on this page assume that the following code has been executed:
+#
+#   require 'pstore'
+#   # Create a store with file +flat.store+.
+#   store = PStore.new('flat.store')
+#   # Store some objects.
+#   store.transaction do
+#     store[:foo] = 0
+#     store[:bar] = 1
+#     store[:baz] = 2
+#   end
+#
+# To avoid modifying the example store, some examples first execute
+# <tt>temp = store.dup</tt>, then apply changes to +temp+.
+#
+# == The Store
+#
+# The contents of the store are maintained in a file whose path is specified
+# when the store is created (see PStore.new).
+# The objects are stored and retrieved using
+# module Marshal, which means that certain objects cannot be added to the store;
+# see {Marshal::dump}[https://docs.ruby-lang.org/en/master/Marshal.html#method-c-dump].
+#
+# == Roots
+#
+# A store may have any number of entries, called _roots_.
+# Each root has a key and a value, just as in a hash:
+#
+# - Key: as in a hash, the key can be (almost) any object;
+#   see {Hash Keys}[https://docs.ruby-lang.org/en/master/Hash.html#class-Hash-label-Hash+Keys].
+#   You may find it convenient to keep it simple by using only
+#   symbols or strings as keys.
+# - Value: the value may be any object that can be marshalled by \Marshal
+#   (see {Marshal::dump}[https://docs.ruby-lang.org/en/master/Marshal.html#method-c-dump])
+#   and in fact may be a collection
+#   (e.g., an array, a hash, a set, a range, etc).
+#   That collection may in turn contain nested objects,
+#   including collections, to any depth;
+#   those objects must also be \Marshal-able.
+#   See {Deep Root Values}[rdoc-ref:PStore@Deep+Root+Values].
+#
+# == Transactions
+#
+# A _transaction_ consists of calls to store-modifying methods (#[]= and #delete)
+# that appear in the block given with a call to method #transaction:
```

Comment:
This seems to imply that only modifications need transactions, but reading also needs transactions.

Reply:
Writes are deferred until the commit;  what about reads?  And what about code not related to pstore?

This passage needs work, regardless or the answers.

Reply:
Well the transaction guarantees that writes won't happen during reads. It also guarantees that reads and writes won't happen when executing code not related to PStore.